libxrdp/libxrdp.c

```diff
@@ -1766,3 +1764,325 @@ libxrdp_send_session_info(struct xrdp_session *session, const char *data,
+                      monitor_index,
+                      monitor_layout->left,
+                      monitor_layout->top,
+                      monitor_layout->right,
+                      monitor_layout->bottom,
+                      monitor_layout->is_primary);
+        }
+        else
+        {
+            /* Per spec (2.2.2.2.1 DISPLAYCONTROL_MONITOR_LAYOUT),
+             * this is the width.
+             * 200 <= width <= 8192 and must not be odd.
+             * Ex: in_uint32_le(s, monitor_layout->width);
+             */
+            in_uint32_le(s, monitor_layout->right);
+            if (monitor_layout->right
+                    > CLIENT_MONITOR_DATA_MAXIMUM_VIRTUAL_MONITOR_WIDTH
+                    || monitor_layout->right
+                    < CLIENT_MONITOR_DATA_MINIMUM_VIRTUAL_MONITOR_WIDTH
+                    || monitor_layout->right % 2 != 0)
+            {
+                return SEC_PROCESS_MONITORS_ERR_INVALID_MONITOR;
+            }
+
+            /* Per spec (2.2.2.2.1 DISPLAYCONTROL_MONITOR_LAYOUT),
+             * this is the height.
+             * 200 <= height <= 8192
+             * Ex: in_uint32_le(s, monitor_layout->height);
+             */
+            in_uint32_le(s, monitor_layout->bottom);
+            if (monitor_layout->bottom
+                    > CLIENT_MONITOR_DATA_MAXIMUM_VIRTUAL_MONITOR_HEIGHT
+                    || monitor_layout->bottom
+                    < CLIENT_MONITOR_DATA_MINIMUM_VIRTUAL_MONITOR_HEIGHT)
+            {
+                return SEC_PROCESS_MONITORS_ERR_INVALID_MONITOR;
+            }
+
+            in_uint32_le(s, monitor_layout->physical_width);
+            in_uint32_le(s, monitor_layout->physical_height);
+
+            in_uint32_le(s, monitor_layout->orientation);
+            switch (monitor_layout->orientation)
+            {
+                case ORIENTATION_LANDSCAPE:
+                case ORIENTATION_PORTRAIT:
+                case ORIENTATION_LANDSCAPE_FLIPPED:
+                case ORIENTATION_PORTRAIT_FLIPPED:
+                    break;
+                default:
+                    LOG(LOG_LEVEL_WARNING, "libxrdp_process_monitor_stream:"
+                        " Orientation is not one of %d, %d, %d, or %d."
+                        " Value was %d and ignored and set to default value of LANDSCAPE.",
+                        ORIENTATION_LANDSCAPE,
+                        ORIENTATION_PORTRAIT,
+                        ORIENTATION_LANDSCAPE_FLIPPED,
+                        ORIENTATION_PORTRAIT_FLIPPED,
+                        monitor_layout->orientation);
+                    monitor_layout->orientation = ORIENTATION_LANDSCAPE;
+            }
+
+            in_uint32_le(s, monitor_layout->desktop_scale_factor);
+            in_uint32_le(s, monitor_layout->device_scale_factor);
+
+            /*
+             * 2.2.2.2.1 DISPLAYCONTROL_MONITOR_LAYOUT
+             */
+            LOG_DEVEL(LOG_LEVEL_TRACE, "libxrdp_process_monitor_stream:"
```

Comment:
I think the level of this log should be promoted. This is informative. What about DEVEL-INFO?

Reply:
Done